### PR TITLE
tikv_util: Support customizing evict policy and operations without promoting for LruCache

### PR DESCRIPTION
**Pull request by MyonKeminta**
### What is changed and how it works?

Issue Number: Ref #11187

What's Changed:

```commit-message
This PR makes the `LruCache` in `tikv_util` to support customizing how to determine an entries in the cache should be evicted.
This is part of solving the issue #11187, which needs a `TxnStatusCache`. The `TxnStatusCache` is desinged to use `LruCache` internally, with ability to get or insert items without promoting items to the head (most-recently-used) position. This PR adds `get_no_promote` and `insert_if_not_exist` functions to `LruCache`.
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

Side effects

\-

### Release note 

```release-note
None
```

### CODE REVIEW

File: components/tikv_util/src/lru.rs [hunk 1]
```diff
@@ -174,14 +174,37 @@ impl<K, V> SizePolicy<K, V> for CountTracker {
     }
 }
 
-pub struct LruCache<K, V, T = CountTracker>
+pub trait GetTailKv<K, V> {
```
Comment (ekexium) [inline components/tikv_util/src/lru.rs:177]: There may be a comment explaining the purpose of the trait. It's unclear from this PR now.

**Review by ekexium — APPROVED** [review]

File: components/tikv_util/src/lru.rs [hunk 2]
```diff
@@ -174,14 +174,37 @@ impl<K, V> SizePolicy<K, V> for CountTracker {
...
+    fn get_tail_kv(&self) -> Option<(&K, &V)>;
+}
+
+pub trait EvictPolicy<K, V> {
```
Comment (cfzjywxk) [inline components/tikv_util/src/lru.rs:181]: Ditto above comments.

File: components/tikv_util/src/lru.rs [hunk 3]
```diff
@@ -292,6 +351,18 @@ where
...
 
+    #[inline]
+    pub fn insert(&mut self, key: K, value: V) {
+        self.insert_impl(key, value, true);
+    }
+
+    /// Insert an entry if the key doesn't exist before. The existing entry
+    /// won't be replaced and won't be promoted to the most-recent place.
+    #[inline]
+    pub fn insert_if_not_exist(&mut self, key: K, value: V) {
```
Comment (cfzjywxk) [inline components/tikv_util/src/lru.rs:362]: How to known if it's inserted successfully, should it return a value?

File: components/tikv_util/src/lru.rs [hunk 4]
```diff
@@ -355,17 +432,37 @@ where
...
+    K: Eq + Hash + Clone + std::fmt::Debug,
+    T: SizePolicy<K, V>,
+    E: EvictPolicy<K, V>,
+{
+    fn get_tail_kv(&self) -> Option<(&K, &V)> {
+        if self.is_empty() {
+            return None;
+        }
+
+        let k = unsafe { self.trace.tail.as_ref().prev.as_ref().key.assume_init_ref() };
```
Comment (zyguan) [inline components/tikv_util/src/lru.rs:446]: maybe extract it as a method of `Trace`?

**Comment by MyonKeminta** [comment]
@ekexium @cfzjywxk @zyguan PTAL again, thanks

**Review by zyguan — APPROVED** [review]

File: components/tikv_util/src/lru.rs [hunk 5]
```diff
@@ -201,6 +242,26 @@ where
...
+    }
+}
+
+impl<K, V, T, E> LruCache<K, V, T, E>
+where
+    T: SizePolicy<K, V>,
+    E: EvictPolicy<K, V>,
+{
+    pub fn new(mut capacity: usize, sample_mask: usize, size_policy: T, evict_policy: E) -> Self {
+        if capacity == 0 {
```
Comment (cfzjywxk) [inline components/tikv_util/src/lru.rs:256]: Better to comment about this.

**Review by cfzjywxk — APPROVED** [review]
LGTM

File: components/tikv_util/src/lru.rs [hunk 6]
```diff
@@ -174,33 +178,82 @@ impl<K, V> SizePolicy<K, V> for CountTracker {
     }
 }
 
-pub struct LruCache<K, V, T = CountTracker>
+/// Some [`EvictPolicy`] may need to know what the entry bing popped out is to
```
Comment (cfzjywxk) [inline components/tikv_util/src/lru.rs:181]: Better do desribe more details about the `Some` here so others could know where is it actually used by now.

Comment (MyonKeminta) [reply components/tikv_util/src/lru.rs:181]: I'll add this description in the next PR.

**Pull request merged**